src/lib/client.js

```diff
@@ -26,18 +30,93 @@ module.exports = class BtpClient {
     this._wsUri = parsedBtpUri.toString().substring('btp+'.length)
   }
 
+  tryToOpenWebSocketClient () {
+    return new Promise((resolve, reject) => {
+      const ws = new WebSocket(this._wsUri, {
+        perMessageDeflate: false
+      })
+      console.log('try to open', this._wsUri)
+      let hasBeenOpen = false
+      ws.on('open', () => {
+        console.log('open!')
+        hasBeenOpen = true
+        resolve(ws)
+      })
+      ws.on('error', (err) => {
+        if (hasBeenOpen) {
+          console.error('WebSocket error', err)
+        } else {
+          reject(err)
+        }
+      })
+      ws.on('close', () => {
+        console.log('close!')
+        // automatically reconnect if server reboots
+        if (hasBeenOpen && !ws.shouldClose) {
+          this.ensureUpstream()
+        }
+      })
+    })
+  }
+
+  getOpenWebSocketClient () {
+    return new Promise((resolve) => {
+      let done = false
+      const tryOnce = () => {
+        this.tryToOpenWebSocketClient().then(ws => {
```

Comment:
Use async/await (since it's already being used in this file)

Reply:
`tryOnce` is a named function called from `setInterval`, and that calls `resolve` in its parent context. I made it an named async function that uses await on line 66, but the resolve still needs to be called from inside the else clause. Also, I needed to add a try-catch clause now, and return from there if the catch clause is reached. All in all I don't think it's an improvement; have a look at the result and let me know if you want me to change it back.